apps/emqx/src/emqx_trace/emqx_trace_freelist.erl

```diff
@@ -0,0 +1,96 @@
+%%--------------------------------------------------------------------
+%% Copyright (c) 2022-2025 EMQ Technologies Co., Ltd. All Rights Reserved.
+%%--------------------------------------------------------------------
+
+-module(emqx_trace_freelist).
+
+-export([
+    %% Constructor:
+    range/2,
+    %% Tracking:
+    occupy/2,
+    %% Accessors:
+    first/1,
+    to_list/1
+]).
+
+-export_type([t/0]).
+
+-opaque t() :: [integer()].
+
+%%
+
+-spec range(integer(), integer()) -> t().
+range(A, B) when A =< B ->
+    [A, B].
+
+-spec first(t()) -> integer() | {error, full}.
+first([A | _]) ->
+    A;
+first([]) ->
+    {error, full}.
+
+-spec to_list(t()) -> [integer()].
+to_list([A, B | Rest]) ->
+    lists:seq(A, B) ++ to_list(Rest);
+to_list([]) ->
+    [].
+
+-spec occupy(integer(), t()) -> t().
```

Comment:
Nit 
a bit confusing that occupying an absent value is not a error, meaning that it's the client of this module should ensure consistency 🤔